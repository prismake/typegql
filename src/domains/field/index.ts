import { fieldsRegistry, FieldInnerConfig } from './registry';

export {
  FieldInnerConfig,
  fieldsRegistry,
  queryFieldsRegistry,
} from './registry';
export { compileAllFields, compileFieldConfig } from './compiler';
export { isQueryField } from './services';
export { FieldError } from './error';

export interface FieldOptions {
  description?: string;
  type?: any;
  name?: string;
  isNullable?: boolean;
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: true,
      ...options,
    };

    const descriptor = Object.getOwnPropertyDescriptor(
      targetInstance,
      fieldName,
    );

    fieldsRegistry.set(targetInstance.constructor, fieldName, {
      isGetter: descriptor && !!descriptor.get,
      ...finalConfig,
    });
  };
}
