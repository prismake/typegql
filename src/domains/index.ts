export {
  ObjectType,
  compileObjectType,
  objectTypeRegistry,
} from './objectType';
export {
  InputObjectType,
  compileInputObjectType,
  inputObjectTypeRegistry,
} from './inputObjectType';
export { DuplexObjectType } from './duplexObjectType';
export { Field } from './field';
export { InputField } from './inputField';
export { DuplexField } from './duplexField';
export { Arg } from './arg';
export { Inject, Context, Source, Info } from './inject';
export { registerEnum, enumsRegistry } from './enum';
export { Union, unionRegistry } from './union';
export { Before, After } from './hooks';
export {
  SchemaRoot,
  Schema,
  schemaRootsRegistry,
  compileSchema,
  Query,
  Mutation,
} from './schema';
