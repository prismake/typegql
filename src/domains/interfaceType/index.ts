import { compileInterfaceTypeWithConfig } from './compiler';
import { interfaceTypeRegistry } from './registry';

export { compileInterfaceType } from './compiler';
export { InterfaceTypeError } from './error';
export { interfaceTypeRegistry } from './registry';

export interface InterfaceTypeOptions {
  name?: string;
  description?: string;
}

export function InterfaceType(options?: InterfaceTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileInterfaceTypeWithConfig(target, config);
    interfaceTypeRegistry.set(target, outputTypeCompiler);
  };
}
