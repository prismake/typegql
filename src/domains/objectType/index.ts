import { compileObjectTypeWithConfig } from './compiler';
import { objectTypeRegistry } from './registry';

export { compileObjectType } from './compiler';
export { ObjectTypeError } from './error';
export { objectTypeRegistry } from './registry';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, config);
    objectTypeRegistry.set(target, outputTypeCompiler);
  };
}
