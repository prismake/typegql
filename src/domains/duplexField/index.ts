import { fieldsRegistry, FieldInnerConfig } from '../field/registry';
import {
  inputFieldsRegistry,
  FieldInputInnerConfig,
} from '../inputField/registry';

export { FieldError } from './error';

export function DuplexField(
  options?: FieldInnerConfig &
    FieldInputInnerConfig & {
      inputNullable?: boolean;
    },
): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    let isNullable = true;
    let inputNullable = true;
    if (options) {
      if (options.isNullable !== undefined) {
        isNullable = options.isNullable;
      }
      if (options.inputNullable !== undefined) {
        inputNullable = options.inputNullable;
      }
      delete options.isNullable;
      delete options.inputNullable;
    }

    const finalInputConfig: FieldInputInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: inputNullable,
      ...options,
    } as any;

    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable,
      ...options,
    } as any;
    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig);

    inputFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      finalInputConfig,
    );
  };
}
