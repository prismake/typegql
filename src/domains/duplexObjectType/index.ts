import { objectTypeRegistry } from '../objectType/registry';
import { compileObjectTypeWithConfig } from '../objectType/compiler';
import { compileInputObjectTypeWithConfig } from '../inputObjectType/compiler';
import { inputObjectTypeRegistry } from '../inputObjectType/registry';
export { ObjectTypeError } from './error';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function DuplexObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const inputConfig = { name: target.name + 'Input', ...options };
    const outputConfig = { name: target.name, ...options };

    const inputTypeCompiler = () =>
      compileInputObjectTypeWithConfig(target, inputConfig);

    inputObjectTypeRegistry.set(target, inputTypeCompiler);

    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, outputConfig);
    objectTypeRegistry.set(target, outputTypeCompiler);
  };
}
