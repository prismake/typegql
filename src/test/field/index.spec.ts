import {
  GraphQLString,
  GraphQLFloat,
  GraphQLBoolean,
  isNamedType,
  getNamedType,
} from 'graphql';
import { ObjectType, Field, compileObjectType } from '~/domains';

import 'reflect-metadata';

describe('Field', () => {
  it('Resolves fields with default value', async () => {
    @ObjectType()
    class Foo {
      @Field() bar: string = 'baz';
      @Field() undef: boolean = undefined;
      @Field() b1: boolean = false;
      @Field() b2: boolean = true;
      @Field() n0: number = 0;
      @Field() nMax: number = Number.MAX_SAFE_INTEGER;
    }
    const compiled = compileObjectType(Foo);

    const barField = compiled.getFields().bar;
    const undefField = compiled.getFields().undef;
    const b1Field = compiled.getFields().b1;
    const b2Field = compiled.getFields().b2;
    const n0Field = compiled.getFields().n0;
    const nMaxField = compiled.getFields().nMax;

    const foo = new Foo();

    expect(await barField.resolve(foo, {}, null, null)).toEqual('baz');
    expect(await undefField.resolve(foo, {}, null, null)).toEqual(undefined);
    expect(await b1Field.resolve(foo, {}, null, null)).toEqual(false);
    expect(await b2Field.resolve(foo, {}, null, null)).toEqual(true);
    expect(await n0Field.resolve(foo, {}, null, null)).toEqual(0);
    expect(await nMaxField.resolve(foo, {}, null, null)).toEqual(
      9007199254740991
    );
  });

  it('Resolves fields with function resolver', async () => {
    @ObjectType()
    class Foo {
      @Field()
      bar(): string {
        return 'baz';
      }
    }

    const compiled = compileObjectType(Foo);
    const barField = compiled.getFields().bar;

    expect(await barField.resolve(new Foo(), {}, null, null as any)).toEqual(
      'baz',
    );
  });

  it('Handles description', () => {
    @ObjectType()
    class Foo {
      @Field({ description: 'test' })
      bar: string = 'baz';
    }
    expect(compileObjectType(Foo).getFields().bar.description).toEqual('test');
  });

  it('Handles custom name', async () => {
    @ObjectType()
    class Foo {
      @Field({ name: 'baz', description: 'test' })
      bar: string = 'test';
    }
    const compiled = compileObjectType(Foo);
    const bazField = compiled.getFields().baz;
    expect(compiled.getFields().bar).toBeFalsy();
    expect(bazField).toBeTruthy();
    expect(bazField.description).toEqual('test');
    expect(await bazField.resolve(new Foo(), {}, null, null as any)).toBe(
      'test',
    );
  });

  it('Properly infers basic scalar types', () => {
    @ObjectType()
    class Foo {
      @Field() bar: string;
      @Field() baz: number;
      @Field() foo: boolean;
      @Field() coo: boolean = false;
      @Field()
      boo(): boolean {
        return true;
      }
    }

    const { bar, baz, foo, boo, coo } = compileObjectType(Foo).getFields();

    expect(bar.type).toEqual(GraphQLString);
    expect(baz.type).toEqual(GraphQLFloat);
    expect(foo.type).toEqual(GraphQLBoolean);
    expect(boo.type).toEqual(GraphQLBoolean);
    expect(coo.type).toEqual(GraphQLBoolean);
  });

  it('Properly sets forced field type', () => {
    @ObjectType()
    class Foo {
      @Field({ type: () => GraphQLFloat })
      bar: string;
    }

    const { bar } = compileObjectType(Foo).getFields();
    expect(bar.type).toEqual(GraphQLFloat);
  });

  it('Supports references to other types', () => {
    @ObjectType()
    class Foo {
      @Field() foo: string;
    }

    @ObjectType()
    class Bar {
      @Field() foo: Foo;
    }

    const { foo } = compileObjectType(Bar).getFields();
    const compiledFoo = compileObjectType(Foo);
    expect(foo.type).toBe(compiledFoo);
  });

  it('Supports references to itself', () => {
    @ObjectType()
    class Foo {
      @Field() fooNested: Foo;
    }

    const { fooNested } = compileObjectType(Foo).getFields();
    expect(fooNested.type).toBe(compileObjectType(Foo));
  });

  it('Supports circular references', () => {
    @ObjectType()
    class Car {
      @Field({ type: () => Owner })
      owner: Owner;
    }

    @ObjectType()
    class Owner {
      @Field({ type: () => Car })
      car: Car;
    }

    const { owner } = compileObjectType(Car).getFields();
    const { car } = compileObjectType(Owner).getFields();

    expect(owner.type).toBe(compileObjectType(Owner));
    expect(car.type).toBe(compileObjectType(Car));
  });

  it('Throws if pointing to unregistered type', () => {
    class Foo {}

    @ObjectType()
    class Bar {
      @Field({ type: () => Foo })
      foo: Foo;
    }

    expect(() =>
      compileObjectType(Bar).getFields(),
    ).toThrowErrorMatchingSnapshot();
  });

  it('Properly resolves native scalar types', () => {
    @ObjectType()
    class Foo {
      @Field({ type: () => String })
      bar: any;
      @Field({ type: () => Number })
      baz: any;
    }

    const { bar, baz } = compileObjectType(Foo).getFields();
    expect(bar.type).toBe(GraphQLString);
    expect(baz.type).toBe(GraphQLFloat);
  });

  it('Shows proper error message when trying to use list type without being explicit about item type', () => {
    @ObjectType()
    class Foo {
      @Field() bar: string[];
    }

    expect(() =>
      compileObjectType(Foo).getFields(),
    ).toThrowErrorMatchingSnapshot();
  });

  it('Shows proper error message when trying to use promise type without being explicit about item type', () => {
    @ObjectType()
    class Foo {
      @Field()
      async bar() {
        return 'baz';
      }
    }

    expect(() =>
      compileObjectType(Foo).getFields(),
    ).toThrowErrorMatchingSnapshot();
  });

  it('Properly supports list type of field', () => {
    @ObjectType()
    class Foo {
      @Field({ type: [String] })
      bar: string[];
    }

    const { bar } = compileObjectType(Foo).getFields();
    expect(isNamedType(bar.type)).toBe(false);
    expect(getNamedType(bar.type)).toBe(GraphQLString);
  });

  it('Is properly passing `this` default values', async () => {
    @ObjectType()
    class Foo {
      private instanceVar = 'instance';
      @Field() bar: string = this.instanceVar;
    }
    const { bar } = compileObjectType(Foo).getFields();
    const resolvedValue = await bar.resolve(new Foo(), null, null, null);
    expect(resolvedValue).toEqual('instance');
  });

  it('Will not allow promise field without type addnotation', async () => {
    @ObjectType()
    class Foo {
      @Field()
      async bar(): Promise<number> {
        return 10;
      }
    }

    expect(() =>
      compileObjectType(Foo).getFields(),
    ).toThrowErrorMatchingSnapshot();
  });
});
