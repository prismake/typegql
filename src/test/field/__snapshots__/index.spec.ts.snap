// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Shows proper error message when trying to use list type without being explicit about item type 1`] = `"@ObjectType Foo.bar: Field returns list so it's required to explicitly set list item type. You can set list type like: @Field({ type: [ItemType] })"`;

exports[`Field Shows proper error message when trying to use promise type without being explicit about item type 1`] = `"@ObjectType Foo.bar: Field returns Promise so it's required to explicitly set resolved type as it's not possible to guess it. You can set resolved type like: @Field({ type: ItemType })"`;

exports[`Field Throws if pointing to unregistered type 1`] = `"@ObjectType Bar.foo: Forced type is incorrect. Make sure to use either native graphql type or class that is registered with @Type decorator"`;

exports[`Field Will not allow promise field without type addnotation 1`] = `"@ObjectType Foo.bar: Field returns Promise so it's required to explicitly set resolved type as it's not possible to guess it. You can set resolved type like: @Field({ type: ItemType })"`;
